src/storage/page_buffer.c

```diff
@@ -7272,56 +7275,63 @@ pgbuf_delete_from_hash_chain (THREAD_ENTRY * thread_p, PGBUF_BCB * bufptr)
 /*
  * pgbuf_lock_page () - Puts a buffer lock on the buffer lock chain
  *   return: If success, PGBUF_LOCK_HOLDER, otherwise PGBUF_LOCK_WAITER
- *   hash_anchor(in):
  *   vpid(in):
+ *   fetch_mode (in):
+ *   hash_anchor(in):
+ *   perf (in/out):
  *
  * Note: This function is invoked only when the page is not in the buffer hash
  *       chain. The caller is holding hash_anchor->hash_mutex.
  *       Before return, the thread releases hash_anchor->hash_mutex.
  */
-static int
-pgbuf_lock_page (THREAD_ENTRY * thread_p, PGBUF_BUFFER_HASH * hash_anchor, const VPID * vpid)
+static PGBUF_LOCK_MODE
+pgbuf_lock_page (THREAD_ENTRY * const thread_p, const VPID * const vpid, PAGE_FETCH_MODE fetch_mode,
+		 PGBUF_BUFFER_HASH * const hash_anchor, PGBUF_FIX_PERF * perf)
 {
+  PGBUF_LOCK_MODE pgbuf_lock_state = PGBUF_LOCK_UNKNOWN;
+
 #if defined(SERVER_MODE)
   PGBUF_BUFFER_LOCK *cur_buffer_lock;
-  THREAD_ENTRY *cur_thrd_entry;
   TSC_TICKS start_tick, end_tick;
   UINT64 lock_wait_time = 0;
 
-  /* the caller is holding hash_anchor->hash_mutex */
-  /* check whether the page is in the Buffer Lock Chain */
-
-  if (thread_p == NULL)
+  // TODO: more accurate comments
+  /* The page is not found in the hash chain the caller is holding hash_anchor->hash_mutex */
+  if (er_errid () == ER_CSS_PTHREAD_MUTEX_TRYLOCK)
```

Comment:
I do not understand what this checks. It seems it expects some error to have been set before , but it does not look right to me that the error being set and checked in different functions.

Reply:
I've had a second look in the code and I have a suggestion. This error check should be moved up to pgbuf_fix.

In pgbuf_fix, the code may be something like this:

```c
  bufptr = pgbuf_search_hash_chain ();
  if (bufptr != nullptr)
    {
      if (pgbuf_bcb_is_direct_victim (bufptr))
        ...
      ...
    }
  else
    {
      if (er_errid () != NO_ERROR)
        {
          // Error searching the hash chain
          pthread_mutex_unlock (&hash_anchor->hash_mutex);
          PGBUF_BCB_CHECK_MUTEX_LEAKS ();
          return nullptr;
        }

      if (fetch_mode == OLD_PAGE_IF_IN_BUFFER)
        ...

      const PGBUF_LOCK_MODE lock_page_res = pgbuf_lock_page (thread_p, vpid, fetch_mode, hash_anchor, &perf);
      ...

      bufptr = pgbuf_claim_bcb_for_fix (...);
      ...
    }
```

As a result, `PGBUF_LOCK_UNKNOWN` may become unnecessary.

Reply:
Done. I would also like to remove UNKNOWN but it is required.